### ISSUE TITLE
Have different rules for in-range updates

### ISSUE THREAD

**Issue by hi-larry** [renovatebot/renovate#1942]
#### This is a:

* [ ] Bug report (non-security related)
* [ ] Feature request
* [x] I'm not sure which of those it is

#### I'm using:

* [ ] The Renovate GitHub App
* [x] Self-hosted GitHub
* [ ] Self-hosted GitLab
* [ ] Self-hosted VSTS

#### Please describe the issue:

I would like to have separate rules for different types of ranged versions and pinned version, similar to how major/minor/patch update type works. I look through the documentation and there seems to be an option for `pin` in which I think is supposed to be pinned version. However, I added `automerge: false` to it but I don't get the expected behaviour:

```
INFO: Creating PR 
       "branchName": "renovate/xxxx-testdep-b-1.x",
       "prTitle": "Update dependency @xxxx/testdep-b to v1.7.0"
INFO: Created Pull Request #24
INFO: PR is configured for automerge
INFO: PR is not ready for merge
```
In short, I have 2 questions/feature requests:

1. It appears that Renovate sees `^1.0.0` and `1.0.0` the same way. Would it be possible to have the variable similar to `isRange` exposed to the configuration?

2. Is there a way to distinguish `^1.0.0` and `~1.0.0` and add different rules to each of them? For example, if the version `1.1.0` is available, I would like `^1.0.0` to be updated and automerged, but NOT automerged `~1.0.0` for minor release.

Many thanks!

**Comment by hi-larry** [comment]
Thank you for explanation, @rarkins 

Let me rephrase it: having the option `upgradeInRange: true` would update `^1.0.0` to `^1.1.0` even though it is part of the range. (Mainly for lock file maintenance) Having a config like:
```
"minor": {
   "automerge": true
}
```
would allow it to be automerged. However, I wouldn't want it to automerge for the change `~1.0.0` to `~1.1.0` because only `~1.0.x` is included in the range. Out of range values should require a review before merging. As well as pinned version, `1.0.0` to `1.1.0`. In my code snippet above, it attempted to automerge.

Therefore, it is important to distinguish all ranged and pinned version. Would my 2 questions be valid reasons for such feature request? If non, is there any workaround for it?

**Comment by rarkins** [comment]
It's not really a matter of ranged vs pinned if you say ^1.0.0 -> ^1.1.0 should be automerged while ~1.0.0 -> ~1.1.0 should not be, as those are both examples of ranges before and after.

It sounds like you want something like `upgradeInRangeAutomerge: true`?

To help me understand better how you're using semver, can you tell me why would you want to use `upgradeInRange`? i.e. what do you gain from using `^1.0.0` and updating it to `^1.1.0` vs using `1.0.0` and `1.1.0` instead?

**Comment by hi-larry** [comment]
The difference between `^1.0.0` and `1.0.0` is that when it is a ranged version, I would like it to be update it automatically, while maintaining the lock file, without any human intervention. e.g `^1.0.0` is the current version and `package-lock.json` has `v1.0.0` locked, I would like it to be both updated to `^1.1.0` and `v1.1.0` respectively. Yes, `v1.1.0` would be part of the ranged values in `^1.0.0` but it won't update `package-lock.json`. For pinned version, it should require a review. So, maybe I would like something like:
```
"minor": {
  "isPin": {
    "automerge": false
  },
  "isRange": {
    "automerge": true
  }
}
```

What do you think of this? Also, `upgradeInRangeAutomerge: true` would also be a good idea for automerging (`^`) and (`~`) values.

**Comment by rarkins** [comment]
Thanks for the additional feedback.

I'm still a little confused though because on one hand you are using pin vs range terminology in your last comment as well as the title of this PR, however on the other hand you mentioned that you wouldn't automerge an upgrade from `~1.0.0` to `~1.1.0` and there's no way we could call that "pinned" - it's from a range to a range. Or am I missing something still?

BTW I assume you've seen Renovate's "lock file maintenance" feature but it doesn't satisfy your requirements to keep lock files updated. I also have in planning to have the ability to update the lock file just like we do with package.json today. e.g. if you have `^1.0.0` in package.json and the version on npmjs goes from `1.0.0` to `1.1.0` then we could raise a PR for it as a "minor" update where only the lock file would get updated and not package.json - i.e. it remains at `^1.0.0`. The main purpose of this would be for library authors who need to maintain wide ranges for compatibility, and who want to use lock files for reproducibility, but who also want the lock file to be regularly updated.

**Comment by hi-larry** [comment]
Thank you for your quick response. That's exactly what I was looking for, hence using the `updateInRange` option. Looking forward to this feature.

Both `^1.0.0` and `~1.0.0` are considered as ranged values, but with different range coverage. `~1.0.0` only covers patch releases, i.e. `~1.0.2` would be within range and shall be automerged, but `~1.1.0` would be out of range. That's what I wasn't sure initially if it was a bug with the option `updateInRange` for (`~`) values. It works perfectly fine with (`^`) values.

To sum up:
- I would like a way to distinguish ranged (`^1.0.0` or `~1.0.0`) and pinned versions (`1.0.0`) and to be able to have options within them just like the update types (major, minor, patch).
- Sorry, I probably didn't explain myself properly earlier but the (`~`) values updates values that are not in range using the option `updateInRange` option. This has nothing to do with range vs pin as it is simply a difference between different types of range values.

**Comment by rarkins** [comment]
Next step is to work out the best way to configure this. While `updateInRangeAutomerge` would satisfy immediate requirements, it seems a shame to invent yet another option that serves such a single purpose and for one person only. So maybe adding a `rangeUpdate` object with `automerge` inside it would be better afterall, if there's possible other uses. Maybe we consider `rangeUpdate` to be an *alternative* update `type` to the existing patch/minor/major.

**Comment by rarkins** [comment]
PinUpdate would be every other type that’s not an in range one

**Comment by hi-larry** [comment]
If I understand correctly, if I had the following config:
```
"minor" {
    "automerge": true
}
"rangeUpdate": {
    "automerge": true
}
```
would the change `~1.0.0` to `~1.1.0` be automerged? My guess is that it would. If so, what should be added to the config to not do so?

**Comment by hi-larry** [comment]
Right, thanks again @rarkins !

**Comment by rarkins** [comment]
You’ve lost me again. Please list maybe 5+ different examples of upgrades and what you do want automerge and do not. For example if you have minor automerge enabled then what do you actually want automerged if you *don’t* want ~1.0.0 to ~1.1.0 to be automerged - which is by an definition a minor update?

**Comment by hi-larry** [comment]
The comment has been removed, only `rangeUpdate` would be enough, thanks!

**Comment by rarkins** [comment]
Waiting on #1954
